File: 304-range-sum-query-2d-immutable/304-range-sum-query-2d-immutable.cpp
typedef vector<int> vi;
typedef vector<vi> vii;


class NumMatrix {
private:
    vii pfx;

public:
    NumMatrix(vector<vector<int>>& matrix) {
        int n = matrix.size();
        int m = matrix[0].size();
        
        pfx = vii(n+1, vi(m+1, 0));
        
        for (int i = 1; i <= n; ++i) {
            for (int j = 1; j <= m; ++j) {
                pfx[i][j] += matrix[i-1][j-1] + pfx[i-1][j] + pfx[i][j-1] - pfx[i-1][j-1];
            }
        }
    }
    
    int sumRegion(int row1, int col1, int row2, int col2) {
        return pfx[row2+1][col2+1] + pfx[row1][col1] - pfx[row2+1][col1] - pfx[row1][col2+1];
    }
};

/**
 * Your NumMatrix object will be instantiated and called as such:
 * NumMatrix* obj = new NumMatrix(matrix);
 * int param_1 = obj->sumRegion(row1,col1,row2,col2);
 */

File: 304-range-sum-query-2d-immutable/304-range-sum-query-2d-immutable.ts
type Row = Array<number>;
type Matrix = Array<Row>;

interface INumMatrix {
    prefixSum: Matrix;
    sumRegion: (row1: number, col1: number, row2: number, col2: number) => number;
}



class NumMatrix implements INumMatrix {
    
    prefixSum: Matrix;

    constructor(matrix: Matrix) {
        
        const n: number = matrix.length;
        const m: number = matrix[0].length;
         
        this.prefixSum = Array.from<Row>({length: n+1});
        
        for (let i = 0; i <= n; ++i) {
            this.prefixSum[i] = Array.from<number>({length: m+1}).fill(0);
        }
        
        for (let i = 0; i < n; ++i) {
            for (let j = 0; j < m; ++j) {
                this.prefixSum[i+1][j+1] = this.prefixSum[i][j+1] + this.prefixSum[i+1][j] - this.prefixSum[i][j] + matrix[i][j];
            }
        }
        
    }

    sumRegion(row1: number, col1: number, row2: number, col2: number): number {
        
        [row2, col2] = [row2 + 1, col2 + 1];

        return (this.prefixSum[row2][col2] - this.prefixSum[row2][col1] - this.prefixSum[row1][col2] + this.prefixSum[row1][col1]);
    }
}

/**
 * Your NumMatrix object will be instantiated and called as such:
 * var obj = new NumMatrix(matrix)
 * var param_1 = obj.sumRegion(row1,col1,row2,col2)
 */